### 데이터 마이닝(Data mining)
축적된 정보 중 가치있는 정보를 발굴하는 과정     
단순 데이터 수집 뿐만 아니라 가공 및 분석과정을 일컫는다.    

#### 데이터 마이닝의 절차
- 문제 정의 : 목표, 요구사항(무슨 데이터는 필요하고,, 무슨 데이터는 필요없고,,), 구현계획     
- 데이터 수집 및 준비 : 매개변수의 선택, 데이터 클리닝 작업    
- 모델 구축 및 평가 : 분석 기법 선택 및 데이터 보정
- 모델 배포 : 새로운 데이터에 모델 적용, 모델 정보 추출 등     

#### 웹 크롤링과 웹 스크래핑
웹 크롤링(Web Crawling)과 웹 스크래핑(Web Scraping)은 구분되는 개념이지만 실무에서는 구분 없이 사용한다.    
- 웹 크롤링 : 웹 크롤러가 다양한 페이지를 인덱싱한다.    
- 웹 스크래핑 : 웹사이트의 데이터를 추출해서 저장한다.      

#### 웹 크롤링
웹 구조에 따른 파싱(parsing) 작업이 이루어지므로 단순 복사-붙여넣기와 다른 특성을 보인다.    
파이썬에서 웹 크롤링 패키지는 Selenium(동적 크롤링 : 사용자와의 상호작용을 포함), BeautifulSoup(정적 크롤링 : 웹페이지의 소스코드를 복사해서 파싱)이 있다.     


In [88]:
#뉴스페이퍼 패키지를 사용해서 뉴스를 크롤링해보자
from newspaper import Article
link = 'https://www.yonhapnewstv.co.kr/news/MYH20240214022300641'
article = Article(link,language='ko')

In [89]:
article.download()
article.parse()
title = article.title
text = article.text
date = article.publish_date

In [90]:
print(title)
print(text[57:])
print(date)

연합뉴스TV

[포인트뉴스] '회삿돈 횡령' 박수홍 친형, 1심 징역 2년…개인자금 사용은 무죄外오늘 각종 사건·사고 소식을 정리해보는, 포인트 뉴습니다.<1> 첫 번째 소식입니다.방송인 박수홍 씨 개인과 회삿돈 수십억 원을 횡령한 혐의로 기소된 친형 진홍 씨가 1심에서 징역 2년을 선고받았습니다.형수에게는 무죄가 선고됐습니다.기소된 금액의 절반 정도만 횡령액으로 인정됐는데, 박 씨 측은 반발했습니다.김유아 기자입니다.<2> 다음 소식입니다.20년 넘게 탈북민을 도와 주목을 받던 목사가 탈북 청소년들을 성추행한 혐의로 실형을 선고받았습니다.법원은 피해자들에게 절대적인 영향력을 미칠 수 있는 지위에서 범행을 저질렀다고 지적했습니다.진기훈 기자입니다.<3> 세 번째 소식입니다.연예인과 전 축구 국가대표 등 유명인들을 내세워 대체불가토큰 'NFT' 프로젝트를 이용해 투자자들에게 피해를 입힌 혐의로 코인업체 대표가 경찰에 입건됐습니다.연루 의혹이 제기된 유명인들은 잇따라 연관성을 부인하고 나섰습니다.한웅희 기자입니다.<4> 마지막 소식입니다.이혼한 뒤 양육비를 주지 않는 부모의 신상을 공개하는 '배드파더스' 운영자가 최근 대법원에서 최종 유죄 판결을 받았습니다.양육자들은 양육비를 받을 별다른 방안이 없다고 호소하고 있는데요.직접 미지급자의 신상을 올리는 '양육비 미투' 캠페인에 나섰습니다.김예린 기자가 취재했습니다.연합뉴스TV 기사문의 및 제보 : 카톡/라인 jebo23(끝)
2024-02-14 13:57:24+00:00


위와 같이 엄청나게 간단하게 크롤링해올 수 있다.     
물론 세세한 크롤링 설정을 할 수 없다는 제약이 있지만 간단하게 제목, 내용, 등을 파악하기엔 가장 편하다

In [91]:
from bs4 import BeautifulSoup
import requests
import pandas as pd

#알라딘 url
page = 200000 #순위별로 페이지 다름
url = f'https://aladin.co.kr/shop/common/wbest.aspx?BestType=Bestseller&BranchType=1&CID=0&page={page}&cnt=1000&SortOrder=1'

req = requests.get(url) #html파일 요청해주는 역할 : request
content = req.content
print(content) #content = html 파일 내용
soup = BeautifulSoup(content,'html.parser')

b'\r\n\r\n<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">\r\n \r\n<html xmlns="http://www.w3.org/1999/xhtml" >\r\n  <head>\r\n    <title>\xec\xa3\xbc\xea\xb0\x84 \xeb\xb2\xa0\xec\x8a\xa4\xed\x8a\xb8  : \xec\x95\x8c\xeb\x9d\xbc\xeb\x94\x98</title>\r\n\r\n\t<meta property="og:type" content="website" />\r\n\t<meta property="og:title" content="\xeb\xb2\xa0\xec\x8a\xa4\xed\x8a\xb8\xec\x85\x80\xeb\x9f\xac : \xec\x95\x8c\xeb\x9d\xbc\xeb\x94\x98" />\r\n\t<meta property="og:url" content="https://www.aladin.co.kr/shop/common/wbest.aspx?BranchType=1&BestType=Bestseller" />\r\n\t<meta property="og:image" content="https://image.aladin.co.kr/img/logo_big.jpg" />\r\n\r\n\t<meta name="twitter:card" content="summary" />\r\n\t<meta name="twitter:title" content="\xeb\xb2\xa0\xec\x8a\xa4\xed\x8a\xb8\xec\x85\x80\xeb\x9f\xac : \xec\x95\x8c\xeb\x9d\xbc\xeb\x94\x98" />\r\n\t<meta name="twitter:url" content="https://www.aladin.co.kr/shop

In [92]:
#html 파일 분석 결과 a.bo3 클래스의 b태그에 책 이름이 들어있다.
#또한 span.ss_p2 의 span에 가격이 들어있다.
book_names = soup.select('a.bo3 b') #a.bo3클래스 b태그
prices = soup.select('span.ss_p2 span') #가격
print(book_names[0].text)
print(int(prices[0].text.replace(',','')))

이처럼 사소한 것들
12420


In [93]:
# 크롤링해서 제목과 가격만 dataframe으로 저장
df = pd.DataFrame(columns=['책제목','가격'])
all_name = []
all_price = []
for i in range(1,4):
    page = i
    url = f'https://aladin.co.kr/shop/common/wbest.aspx?BestType=Bestseller&BranchType=1&CID=0&page={page}&cnt=1000&SortOrder=1'
    req = requests.get(url)
    content = req.content
    soup = BeautifulSoup(content,'html.parser')
    book_names = soup.select('a.bo3 b') #a.bo3클래스 b태그
    prices = soup.select('span.ss_p2 span') #가격
    length = len(book_names)
    for j in range(length):

        all_name.append(book_names[j].text)
        all_price.append(int(prices[j].text.replace(',','')))
leng = len(all_name)
df['책제목'] = [all_name[j] for j in range(leng)]
df['가격'] = [all_price[j] for j in range(leng)]

print(df)
df.to_csv('./알라딘크롤링결과.csv')

                              책제목     가격
0                      이처럼 사소한 것들  12420
1                   아이는 무엇으로 자라는가  15750
2                        너를 위한 삼월  19800
3           나는 메트로폴리탄 미술관의 경비원입니다  15750
4                    마흔에 읽는 쇼펜하우어  15300
..                            ...    ...
145                            자살  14400
146  부자 아빠 가난한 아빠 1 (20주년 특별 기념판)  15300
147                           랑데부  18450
148                       AI 2024  19800
149            꼬마비 만화 전집 세트 - 전5권  99000

[150 rows x 2 columns]


실습에서는 페이지 하나만 크롤링했지만 나는 150위까지 크롤링을 해보고 싶어서 코드를 변형해서 작성했다.     
살짝 복잡해보일 수 있지만, 전체 데이터를 모아놨다가 data frame에 입히는 코드이다.

In [94]:
#spans.u_cbox_text_contents
#em.u_cbox_cnt_recomm
#더보기 span.u_cbox_page_more

### 동적 페이지에서 크롤링하기
알라딘 중고서적과 같이 하나의 페이지 당 하나의 주소로 구성되어있고 변하지 않는 정적인 페이지는 Beautifulsoup을 사용하면 간단하게 크롤링할 수 있었다.      
하지만 최근 페이지들은 대부분 반응형, 즉 사용자와 상호작용하면서 구성되어있는 동적인 페이지가 많다.     
이런 동적인 페이지에서 크롤링을 하기 위해서는 **Selenium**을 사용한다.     
이번에는 네이버 지도에서 크롤링하는 것과 네이버 뉴스 기사 댓글(더보기를 통해서 계속 불러올 수 있는)을 크롤링해보도록 하자

In [95]:
import pandas as pd
from selenium import webdriver #크롬창을 조종하기 위함
from selenium.webdriver.common.by import By #웹사이트의 구성요소를 선택하기 위함
from selenium.webdriver.support.ui import WebDriverWait #웹페이지가 전부 로드될때까지 기다리기 위함
from webdriver_manager.chrome import ChromeDriverManager #크롬에서 크롤링을 하기 위해서 웹드라이버를 설치하는 모듈
from selenium.webdriver.support import expected_conditions as EC #크롬에서 어떤 부분의 상태를 확인하는 모듈
import time

In [129]:
driver = webdriver.Chrome() #웹 드라이버를 설치하고 조종할 수 있는 크롬창 오픈

In [130]:
#네이버 지도에서 원하는 키워드를 검색하자
url = 'https://map.naver.com/v5/search/서울 노원구 카페'
driver.get(url) #나는 서울 노원구 카페를 검색어로 사용할 것이다.

- 위 코드를 실행하면 열려있던 조종되는 크롬창에서 서울 노원구 카페를 검색한 것을 확인할 수 있다.
- 즉 get이 url을 받아서 검색을 해준다고 보면 된다. 이 기능은 다른 홈페이지로 이동할때도 요긴하게 사용된다.
- 우리가 웹페이지에 접속할 때 인터넷 회선, 서버 문제로 접속시간이 소요된다.
- 이런 문제를 고려하지 않고 코드를 바로 실행시켜버리면 문제가 발생한다.
- 문제를 해결하기 위해서 로딩이 다 끝날때까지 기다리는 나름 무식한 방법을 사용한다 ㅎ
- Implicity wait와 Explicity를 사용해서 로딩을 기다리는데, 전자는 정해진 시간만큼 기다리는 방법이고, 시간을 넉넉하게 설정하면 에러가 발생하지 않아서 사람이 신경쓸 일이 별로 없다.
- 후자의 방법은 인터넷 상태가 변동이 클때나 특정한 구성요소가 로딩되는 즉시 코드를 실행할 때 사용된다.
- 지도의 경우 검색창만 다 로딩되면 되니까 Explicity 방법을 사용할 수 있다.

In [131]:
try:
    element = WebDriverWait(driver,10).until(
        EC.presence_of_element_located((By.CLASS_NAME,"input_search"))
    ) #네이버 지도의 검색창은 'input_search'라는 클래스 이름으로 설정되어있다.
finally:
    pass

- 네이버 지도는 특정 영역을 선택하지 않을 경우에 셀레니움 크롤링 기능이 적용되지 않는다. 
- 따라서 클릭 또는 복사할 영역을 한번 지정해줘야한다. driver.switch_to.frame / driver.find_elements를 적절하게 사용한다.
- 네이버 지도의 검색결과에서 모든 페이지를 크롤링하기 위해서 반복문을 이용한다.
- 마지막 페이지가 어디인지 정확하게 모르니까 새로운 정보가 나오지 않으면 탈출하는 방식 사용


In [99]:
'''
driver.switch_to.frame("searchIframe") #searchIframe은 네이버 지도에서 검색결과가 나타나는 iframe id이다.

firstname = '' #맨 첫번째 상호 이름. 다음페이지에서도 똑같은 상호가 나오면 다음 페이지가 없다고 판단
brk = 1
res = pd.DataFrame() #결과 저장할 데이터프레임 생성
empty = '//*[@id="_pcmap_list_scroll_container"]' #크롤링 할 데이터가 있는 searIframe에서 빈 공간을 입력함

while brk:
    driver.find_element(By.XPATH,empty) #XPATH기준으로 ""를 찾는다. XPATH는 마크업 언어에서 특정 요소를 찾기위한 경로이다.
    #보통은 id나 클래스 이름으로 요소를 찾는데 XPATH를 사용하면 @id="_pcmap_list_scroll_container"이 있는 곳을 빠르게 찾을 수 있다.
    for i in range(1,51): #1-50번째 상호 순회
        nm = ['NA'] #상호가 저장될 변수
        addr = ['NA'] #주소가 저장될 변수
        driver.find_element(By.XPATH,empty)
        # 네이버 지도는 이름이 두 가지 영역에 저장되는 것을 확인했습니다. 일단 둘 다 긁어오고, 이름이 있는지는 나중에 체크하도록 했습니다
        nm = driver.find_elements(By.XPATH, f'//*[@id="_pcmap_list_scroll_container"]/ul/li[{i}]/div[1]/div[2]/a[1]/div/div/span[1]')
        nm += driver.find_elements(By.XPATH, f'//*[@id="_pcmap_list_scroll_container"]/ul/li[{i}]/div[1]/div/a[1]/div/div/span[1]')
        if nm!=[]:
            res = pd.concat([res,pd.DataFrame([nm[0].text,addr]).T])
            res.to_csv('./nowon_cafe.csv',index=False)
        if i == 1: #첫번째 불러왔으면 이전페이지의 상호와 같은지 확인 후 같다면 탈출
            if firstname == nm:
                brk = 0
                break
            else:
                firstname = nm


# 다음 페이지로 넘어가는 코드입니다 다음 버튼을 인식해서 클릭하도록 만들었습니다
driver.find_element(By.CLASS_NAME,'zRM9F')
driver.find_element(By.XPATH, 'place_blind').click()
#driver.find_element(By.XPATH, '//*[@id="app-root"]/div/div[2]/div[2]/a[7]').click()
time.sleep(2) # 페이지 로딩 시간 2초
'''

'\ndriver.switch_to.frame("searchIframe") #searchIframe은 네이버 지도에서 검색결과가 나타나는 iframe id이다.\n\nfirstname = \'\' #맨 첫번째 상호 이름. 다음페이지에서도 똑같은 상호가 나오면 다음 페이지가 없다고 판단\nbrk = 1\nres = pd.DataFrame() #결과 저장할 데이터프레임 생성\nempty = \'//*[@id="_pcmap_list_scroll_container"]\' #크롤링 할 데이터가 있는 searIframe에서 빈 공간을 입력함\n\nwhile brk:\n    driver.find_element(By.XPATH,empty) #XPATH기준으로 ""를 찾는다. XPATH는 마크업 언어에서 특정 요소를 찾기위한 경로이다.\n    #보통은 id나 클래스 이름으로 요소를 찾는데 XPATH를 사용하면 @id="_pcmap_list_scroll_container"이 있는 곳을 빠르게 찾을 수 있다.\n    for i in range(1,51): #1-50번째 상호 순회\n        nm = [\'NA\'] #상호가 저장될 변수\n        addr = [\'NA\'] #주소가 저장될 변수\n        driver.find_element(By.XPATH,empty)\n        # 네이버 지도는 이름이 두 가지 영역에 저장되는 것을 확인했습니다. 일단 둘 다 긁어오고, 이름이 있는지는 나중에 체크하도록 했습니다\n        nm = driver.find_elements(By.XPATH, f\'//*[@id="_pcmap_list_scroll_container"]/ul/li[{i}]/div[1]/div[2]/a[1]/div/div/span[1]\')\n        nm += driver.find_elements(By.XPATH, f\'//*[@id="_pcmap_list_scroll_container"]/ul/li[{i}]/div[1]/div/a[

In [132]:
driver.switch_to.frame("searchIframe") #searchIframe은 네이버 지도에서 검색결과가 나타나는 iframe id이다.

firstname = '' #맨 첫번째 상호 이름. 다음페이지에서도 똑같은 상호가 나오면 다음 페이지가 없다고 판단
brk = 1
res = pd.DataFrame() #결과 저장할 데이터프레임 생성
empty = '//*[@id="_pcmap_list_scroll_container"]' #크롤링 할 데이터가 있는 searIframe에서의 공간


In [134]:
scroll_place = driver.find_element(By.ID,'_pcmap_list_scroll_container')
last_height = driver.execute_script("return document.body.scrollHeight",scroll_place) 
from bs4 import BeautifulSoup
while brk:
    
    driver.execute_script("arguments[0].scrollTop = arguments[0].scrollHeight", scroll_place)    
    
    scroll_height = driver.execute_script("return arguments[0].scrollHeight", scroll_place)
    if last_height == scroll_height:
        print("Reached the bottom of the scrollable element.")
        raw = driver.find_element(By.XPATH,empty)
        print(raw.text)
        nm = driver.find_elements(By.XPATH, f'//*[@id="_pcmap_list_scroll_container"]/ul/li[{0}]/div[1]/div[2]/a[1]/div/div/span[1]')
        print(nm)
        break
    else:
        last_height = scroll_height

    time.sleep(2)

Reached the bottom of the scrollable element.
마스트리스보트카페,디저트
오늘 휴무
리뷰 231
서울 노원구 공릉동
상세주소 열기
저장
마스트리스보트
마스트리스보트
마스트리스보트
웃을때보조개콕
마스트리스보트
coffe hyang
고양이가 귀여워요 플랫화이트, 아아, 시나몬애플티를 주문했고 다들 만족했습니다
우연히 의정부 왔다가 들렸어요 귀여운 고양이 그리고 맛난 까지 종종 올게요
리프노즈드아쉬지카페,디저트
영업 중리뷰 497
TV
놀라운토요일
저장
리프노즈드아쉬지
리프노즈드아쉬지
리프노즈드아쉬지
내가왕초
리프노즈드아쉬지
미정왕
제가 많이 애정하는 디저트가게예요 사장님도 친절하시고 디저트도 종류가 많고 맛있어요
공릉에서 가장 맛있는 곳 얼쌈도 취향
리브인오후베이커리
영업 중
리뷰 580
서울 노원구 공릉동
상세주소 열기
저장
리브인오후
리브인오후
리브인오후
차차랑밍밍
리브인오후
디저트러버 영이
공릉동 케이크 맛집입니다. 구황작물 케이크는 할매 입맛분들 진짜 좋아하실 것 같은 맛이에요.
특별한 자리에 필요해서 모듬으로 구성된 홀를 예약했습니다.
감각카페
영업 중
별점
4.5리뷰 999+
저장
감각
감각
감각
아기복어18
감각
WonGyeong
수플레로 유명한 집이지만 케이크도 짱 맛있었유 커피랑 넘넘 잘 어울리고 대화하기 조아요
말렌카 좋아해서 가끔 보이면 사먹는데 딱 있더라구요 ㅎㅎ 위에 크림 얹어주시구요 ㅎㅎ
라라브레드 공릉점
네이버톡톡
베이커리
영업 중
별점
4.44리뷰 999+
TV
생방송오늘아침
저장
라라브레드 공릉점
라라브레드 공릉점
라라브레드 공릉점
건강큐
라라브레드 공릉점
빙벵봉11
먹고 싶은 빵이 너무 많아요. 모카소금빵 맛있고, 팡도르도 사르르 녹네요.
올리브 치아바타랑 한잔 했는데 넘 맛있어서 몇개 더 포장해서 왔네요ㅋ 가격도 좋습니다
백오얏카페
영업 중
리뷰 321
서울 노원구 상계동
상세주소 열기
저장
백오얏
백오얏
백오얏
맛있는데만가는빵덕
백오얏
먹제리
바스크치즈케이크랑 땅콩크림라떼, 아인슈페너 먹었는데 

In [123]:
print(scroll_height)

13755
